File: code/js/src/services/utils/siren/SirenEntity.ts
import {Link} from "./Link";
import {Action} from "./Action";
import {SubEntity} from "./SubEntity";

/**
 * Siren is a specification for representing hypermedia entities in JSON.
 *
 * @see <a href="https://github.com/kevinswiber/siren">Siren Specification</a>
 *
 * @property class the class of the entity (optional)
 * @property properties the properties of the entity (optional)
 * @property entities the sub-entities of the entity (optional)
 * @property actions the actions that can be performed on the entity (optional)
 * @property links the links to other entities (optional)
 * @property title the title of the entity (optional)
 */
export interface SirenEntity<T> {
    class?: string[]
    properties?: T
    entities?: SubEntity<unknown>[]
    actions?: Action[]
    links?: Link[]
    title?: string
}

/**
 * Siren is a specification for representing hypermedia entities in JSON.
 *
 * @see <a href="https://github.com/kevinswiber/siren">Siren Specification</a>
 *
 * @property class the class of the entity (optional)
 * @property properties the properties of the entity (optional)
 * @property entities the sub-entities of the entity (optional)
 * @property actions the actions that can be performed on the entity (optional)
 * @property links the links to other entities (optional)
 * @property title the title of the entity (optional)
 */
export class SirenEntity<T> {
    constructor(entity: SirenEntity<T>) {
        this.class = entity.class;
        this.properties = entity.properties;
        this.entities = entity.entities;
        this.actions = entity.actions;
        this.links = entity.links;
        this.title = entity.title
    }
}

export const sirenMediaType = "application/vnd.siren+json";

File: code/jvm/src/main/kotlin/pt/isel/daw/battleships/http/media/siren/SirenEntity.kt
package pt.isel.daw.battleships.http.media.siren

import org.springframework.http.MediaType

/**
 * Siren is a specification for representing hypermedia entities in JSON.
 *
 * @see <a href="https://github.com/kevinswiber/siren">Siren Specification</a>
 *
 * @property class the class of the entity (optional)
 * @property properties the properties of the entity (optional)
 * @property entities the sub-entities of the entity (optional)
 * @property actions the actions that can be performed on the entity (optional)
 * @property links the links to other entities (optional)
 * @property title the title of the entity (optional)
 */
data class SirenEntity<T>(
    val `class`: List<String>? = null,
    val properties: T? = null,
    val entities: List<SubEntity>? = null,
    val actions: List<Action>? = null,
    val links: List<Link>? = null,
    val title: String? = null
) {
    companion object {
        private const val APPLICATION_TYPE = "application"
        private const val SIREN_SUBTYPE = "vnd.siren+json"
        const val SIREN_MEDIA_TYPE = "$APPLICATION_TYPE/$SIREN_SUBTYPE"

        @Suppress("unused")
        val sirenMediaType = MediaType(APPLICATION_TYPE, SIREN_SUBTYPE)
    }
}
